paddle/operators/add_op.cc

```diff
@@ -21,17 +21,19 @@ namespace operators {
 
 class AddOp : public framework::OperatorWithKernel {
 protected:
-  void InferShape(
-      const std::vector<const framework::Tensor *> &inputs,
-      const std::vector<framework::Tensor *> &outputs) const override {
-    PADDLE_ENFORCE(inputs.size() == 2, "Input size of AddOp must be two");
-    PADDLE_ENFORCE(outputs.size() == 1, "Output size of AddOp must be one");
-    PADDLE_ENFORCE(
-        inputs[0] != nullptr && inputs[1] != nullptr && outputs[0] != nullptr,
-        "Inputs/Outputs of AddOp must all be set");
-    PADDLE_ENFORCE(inputs[0]->dims() == inputs[1]->dims(),
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE(ctx.InputSize() == 2, "Input size of AddOp must be two");
+    PADDLE_ENFORCE(ctx.OutputSize() == 1, "Output size of AddOp must be one");
+    PADDLE_ENFORCE(ctx.Input<framework::Variable>(0) != nullptr &&
```

Comment:
currently, the return type is a pointer, look like this

```c++
template <typename T = framework::Variable>
T* Input(size_t idx);
```

but this does not support `std:: string`, that means we will add another 4 interfaces like:

```c++
std::string InputName();
std::vector<std::string> MultiInputNames();
std::string OutputName();
std::vector<std::string> MultiOutputNames();
```
Is there some way to embed `std::string` as a template type as follows ?
```c++
auto variable_name = Input<std::string>("x") 
```
@jacquesqiao @reyoung